### ISSUE TITLE
Why can't  I get a second order derivative of a loss function containing a NN w.r.t. its parameters?

### ISSUE THREAD

**Issue by stash-196** [FluxML/Flux.jl#1518]
Ultimately, I want to make a function that calculates the hessian matrix of a function containing a neural network w.r.t. its parameters, like `hessian_wrt_all_params(function, model)` from [this](https://github.com/stash-196/pytorch-derivatives/blob/main/derivatives.py)

Consider the example below
```julia
using Flux: Chain, Dense, σ, crossentropy, params

model = Chain(
    x -> reshape(x, :, size(x, 4)),
    Dense(2, 5),
    Dense(5, 1),
    x -> σ.(x)
)
n_data = 5
input = randn(2, 1, 1, n_data)
target = randn(1, n_data)
loss = model -> crossentropy(model(input), target)
```

I started with the most naive approach, which is to calculate the derivatives element-wise using `gradient`, but I can not.

```julia
flgrad = model -> Flux.gradient(loss, model)[1][1][2][1][1]
flgrad(model)
flgradgrad =Flux.gradient(flgrad, model)    # ERROR: Can't differentiate foreigncall expression
```
Why doesn't this work?

**Comment by ToucheSir** [comment]
If you want a hessian, have you tried using https://fluxml.ai/Zygote.jl/latest/utils/#Zygote.hessian? Many functions will not work with nested Zygote reverse mode (this is a known limitation).

**Comment by CarloLucibello** [comment]
Closing as a support question that was answered: nested Zygote reverse-over-reverse has known limitations, and `Zygote.hessian` (forward-over-reverse) is the supported route for second derivatives.

Since this was filed, `Flux.gradient` also supports Enzyme (via `Duplicated`) and Mooncake through ADTypes, which handle much more than nested Zygote. Discourse is the better venue for follow-up usage questions.
